lib/Dialect/FIRRTL/Transforms/BlackboxMemory.cpp

```diff
@@ -0,0 +1,184 @@
+    memName = op.name().getValue();
+  auto nameAttr = builder.getStringAttr(memName);
+
+  // Get the memory port descriptors.  This gives us the name and kind of each
+  // memory port created by the MemOp.
+  SmallVector<std::pair<Identifier, MemOp::PortKind>, 2> memPorts;
+  op.getPorts(memPorts);
+
+  // Create an external module for this blackbox memory. This external module
+  // must be compatible with the modules which are generated for `vlsi_mem_gen`,
+  // which can be found in the rocketchip project.
+  // The external module port list, 1R + 1W memory has 4+5=9 fields.
+  SmallVector<ModulePortInfo, 9> extPortList;
+  unsigned readPorts = 0;
+  unsigned writePorts = 0;
+  unsigned readWritePorts = 0;
+  for (unsigned i = 0; i < memPorts.size(); ++i) {
+    // Calculate the naming prefix to use based on the kind of the port.
+    std::string prefix;
+    switch (memPorts[i].second) {
+    case MemOp::PortKind::Read:
+      prefix = (Twine("R") + Twine(readPorts++) + "_").str();
+      break;
+    case MemOp::PortKind::Write:
+      prefix = (Twine("W") + Twine(writePorts++) + "_").str();
+      break;
+    case MemOp::PortKind::ReadWrite:
+      prefix = (Twine("RW") + Twine(readWritePorts++) + "_").str();
+      break;
+    }
+    // Flatten the bundle representing a memory port, name-mangling and adding
+    // every field in the bundle to the exter module's port list.
+    auto type = op.getResult(i).getType().cast<FIRRTLType>();
+    for (auto field : type.getPassiveType().cast<BundleType>().getElements()) {
+      auto fieldType = SubfieldOp::getResultType(type, field.name, op.getLoc());
+      extPortList.push_back({builder.getStringAttr(prefix + field.name.str()),
+                             FlipType::get(fieldType)});
+    }
+  }
+
+  // Create the blackbox external module.  The module's name is the name of the
+  // memory postfixed with "_ext".
+  std::string extName = memName.str() + "_ext";
+  auto extModule = builder.create<FExtModuleOp>(
+      op.getLoc(), builder.getStringAttr(extName), extPortList, extName);
+
+  // Add the memory parameters as attributes on the external module.  These will
+  // be used by the generator tools to create memory configuration files, or
+  // create the memory itself. For now these are added directly to the external
+  // module's attribute dictionary, but in the future they should be added in a
+  // way which is made available to generators.  We use a named attr list to
+  // avoid creating lots of intermediate dictionary attributes.
+  NamedAttrList attrs(extModule->getAttrDictionary());
+  attrs.set("generator", builder.getStringAttr("FIRRTLMemory"));
+  attrs.set("readLatency", op.readLatencyAttr());
+  attrs.set("writeLatency", op.writeLatencyAttr());
+  attrs.set("depth", op.depthAttr());
+  attrs.set("ruw", op.ruwAttr());
+  extModule->setAttrs(attrs.getDictionary(builder.getContext()));
+
+  // Create a regular module to wrap the external module, instantiating the
```

Comment:
I understand doing this to be compatible with the scala firrtl lowering, but should it be optional?  What purpose does it serve?

Reply:
It is now optional, and the pass now has an option `emit-wrappers`.  Implementing this required quite a large duplication of logic.  I do not know what the original purpose of the wrapper module was.